File: csharp-playwright-framework/PlaywrightFramework/Utilities/Logger.cs
using Serilog;

namespace PlaywrightFramework.Utilities;

/// <summary>
/// Custom logger for test execution
/// </summary>
public static class TestLogger
{
    static TestLogger()
    {
        Log.Logger = new LoggerConfiguration()
            .MinimumLevel.Debug()
            .WriteTo.Console()
            .WriteTo.File("logs/test-.log", rollingInterval: RollingInterval.Day)
            .CreateLogger();
    }

    public static void Info(string message) => Log.Information("ℹ️  {Message}", message);
    public static void Debug(string message) => Log.Debug("🔍 {Message}", message);
    public static void Warning(string message) => Log.Warning("⚠️  {Message}", message);
    public static void Error(string message) => Log.Error("❌ {Message}", message);
    public static void Error(Exception ex, string message) => Log.Error(ex, "❌ {Message}", message);
    public static void Success(string message) => Log.Information("✅ {Message}", message);
    public static void TestStart(string testName) => Log.Information("🧪 Starting test: {TestName}", testName);
    public static void TestEnd(string testName, bool passed)
    {
        if (passed)
            Log.Information("✅ Test passed: {TestName}", testName);
        else
            Log.Error("❌ Test failed: {TestName}", testName);
    }
    public static void Step(string step) => Log.Information("📝 Step: {Step}", step);
    public static void Api(string method, string endpoint) => Log.Information("🌐 API: {Method} {Endpoint}", method, endpoint);

    public static void CloseAndFlush() => Log.CloseAndFlush();
}

File: python-playwright-framework/utils/logger.py
"""
Logger utility for test execution logging
Provides structured logging with different levels
"""

import logging
import os
from pathlib import Path
from datetime import datetime


class Logger:
    """Custom logger for test automation"""
    
    def __init__(self, log_file_name: str = "test_execution.log", console_output: bool = True):
        # Create logs directory
        logs_dir = Path(os.getcwd()) / "reports" / "logs"
        logs_dir.mkdir(parents=True, exist_ok=True)
        
        self.log_file = logs_dir / log_file_name
        self.console_output = console_output
        
        # Setup logger
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.DEBUG)
        
        # Remove existing handlers
        self.logger.handlers = []
        
        # File handler
        file_handler = logging.FileHandler(self.log_file, mode='w')
        file_handler.setLevel(logging.DEBUG)
        file_formatter = logging.Formatter(
            '%(asctime)s - %(levelname)s - %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S'
        )
        file_handler.setFormatter(file_formatter)
        self.logger.addHandler(file_handler)
        
        # Console handler
        if console_output:
            console_handler = logging.StreamHandler()
            console_handler.setLevel(logging.INFO)
            console_formatter = logging.Formatter(
                '%(levelname)s - %(message)s'
            )
            console_handler.setFormatter(console_formatter)
            self.logger.addHandler(console_handler)
    
    def _get_emoji(self, level: str) -> str:
        """Get emoji for log level"""
        emojis = {
            'INFO': 'ℹ️',
            'WARNING': '⚠️',
            'ERROR': '❌',
            'DEBUG': '🐛',
            'SUCCESS': '✅',
            'START': '🚀',
            'END': '🏁'
        }
        return emojis.get(level, '')
    
    def info(self, message: str) -> None:
        """Log info message"""
        self.logger.info(f"{self._get_emoji('INFO')} {message}")
    
    def warning(self, message: str) -> None:
        """Log warning message"""
        self.logger.warning(f"{self._get_emoji('WARNING')} {message}")
    
    def error(self, message: str) -> None:
        """Log error message"""
        self.logger.error(f"{self._get_emoji('ERROR')} {message}")
    
    def debug(self, message: str) -> None:
        """Log debug message"""
        self.logger.debug(f"{self._get_emoji('DEBUG')} {message}")
    
    def success(self, message: str) -> None:
        """Log success message"""
        self.logger.info(f"{self._get_emoji('SUCCESS')} {message}")
    
    def test_start(self, test_name: str) -> None:
        """Log test start"""
        self.logger.info(f"{self._get_emoji('START')} Test Started: {test_name}")
    
    def test_end(self, test_name: str, status: str) -> None:
        """Log test end"""
        emoji = '✅' if status == 'PASSED' else '❌'
        self.logger.info(f"{emoji} Test {status}: {test_name}")
    
    def step(self, step_description: str) -> None:
        """Log step in test"""
        self.logger.info(f"  📌 Step: {step_description}")
    
    def api(self, method: str, endpoint: str, status: int, duration: float) -> None:
        """Log API request/response"""
        self.logger.info(
            f"🌐 API {method} {endpoint} - Status: {status} - Duration: {duration:.2f}ms"
        )


# Singleton instance
logger = Logger()
